diff --git a/vendor/pulp-platform/common_cells/src/popcount.sv b/vendor/pulp-platform/common_cells/src/popcount.sv
index 72b9b71f0..6fde114f3 100644
--- a/vendor/pulp-platform/common_cells/src/popcount.sv
+++ b/vendor/pulp-platform/common_cells/src/popcount.sv
@@ -30,10 +30,7 @@ module popcount #(
    logic [PopcountWidth-2:0]         left_child_result, right_child_result;
 
    //Zero pad the input to next power of two
-   always_comb begin
-     padded_input = '0;
-     padded_input[INPUT_WIDTH-1:0] = data_i;
-   end
+   assign padded_input = {{{PaddedWidth-INPUT_WIDTH}{1'b0}}, data_i};
 
    //Recursive instantiation to build binary adder tree
    if (INPUT_WIDTH == 1) begin : single_node
