diff --git a/scripts/lib.py b/scripts/lib.py
index ef9744d..828a3ea 100644
--- a/scripts/lib.py
+++ b/scripts/lib.py
@@ -227,7 +227,11 @@ def process_regression_list(testlist, test, iterations, matched_list,
         "Processing regression test list : {}, test: {}".format(testlist, test))
     yaml_data = read_yaml(testlist)
     mult_test = test.split(',')
-    for entry in yaml_data:
+    if 'testlist' in yaml_data and isinstance(yaml_data['testlist'], list):
+        yaml_testlist = yaml_data['testlist']
+    else:
+        yaml_testlist = yaml_data
+    for entry in yaml_testlist:
         if 'import' in entry:
             sub_list = re.sub('<riscv_dv_root>', riscv_dv_root, entry['import'])
             process_regression_list(sub_list, test, iterations, matched_list,
