diff --git a/arch_test_target/spike/model_test.h b/arch_test_target/spike/model_test.h
index e968e43a..7628af51 100644
--- a/arch_test_target/spike/model_test.h
+++ b/arch_test_target/spike/model_test.h
@@ -23,6 +23,7 @@
   li x1, 1;                                                                   \
   write_tohost:                                                               \
     sw x1, tohost, t1;                                                        \
+  ecall
   self_loop:  j self_loop;
 
 #define RVMODEL_BOOT
-- 
2.39.0

