diff --git a/output/Makefile b/output/Makefile
index cf1214f..c81bccc 100644
--- a/output/Makefile
+++ b/output/Makefile
@@ -20,9 +20,9 @@ extra_files =
 #--------------------------------------------------------------------

 RISCV_GCC = riscv64-unknown-elf-gcc
-RISCV_GCC_OPTS = -nostdlib -nostartfiles -Wa,-march=RVIMAFDXhwacha
+RISCV_GCC_OPTS = -nostdlib -nostartfiles -Wa,-march=rv64imc
 RISCV_OBJDUMP = riscv64-unknown-elf-objdump --disassemble-all --section=.text --section=.data --section=.bss
-RISCV_SIM = spike --extension=hwacha
+RISCV_SIM = spike

 #------------------------------------------------------------
 # Build assembly tests
@@ -38,9 +38,6 @@ $(asm_tests_dump): %.dump: %
 $(asm_tests_bin): %: %.S $(extra_files)
 	$(RISCV_GCC) $(RISCV_GCC_OPTS) -I../env/p -T../env/p/link.ld $< -o $@

-$(asm_tests_hex): %.hex: % $(extra_files)
-	elf2hex 16 16384 $< > $@
-
 $(asm_tests_sig): %.sig: %
 	$(RISCV_SIM) +signature=$@ $<

@@ -51,12 +48,12 @@ run: $(asm_tests_sig)
 	echo; perl -ne 'print "  [$$1] $$ARGV \t$$2\n" if /\*{3}(.{8})\*{3}(.*)/' \
 	       $(asm_tests_sig); echo;

-junk += $(asm_tests_bin) $(asm_tests_dump) $(asm_tests_sig) $(asm_tests_hex)
+junk += $(asm_tests_bin) $(asm_tests_dump) $(asm_tests_sig)

 #------------------------------------------------------------
 # Default

-all: $(asm_tests_dump) $(asm_tests_hex)
+all: $(asm_tests_dump)

 #------------------------------------------------------------
 # Clean up
